src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+import chisel3._
+import chisel3.experimental.SourceInfo
+import chisel3.reflect.DataMirror
+import chisel3.util.{log2Up, BitPat, Mux1H, PriorityMux}
+import chisel3.util.experimental.decode.{decoder, TruthTable}
+import chisel3.util.experimental.InlineInstance
+import scala.collection.SeqMap
+import scala.collection.mutable.{ArrayBuffer, LinkedHashMap, LinkedHashSet}
+
+/** A sparse vector.  Under the hood, this is a [[Record]] that provides a
+  * dynamic index as if it were a [[Vec]].
+  *
+  * [[SparseVec]] implements a number of dynamic index edge case behaviors to
+  * make it behave exactly the same as a dense [[Vec]] that was sparsely
+  * initialized (and following legacy conventions of how FIRRTL compilers have
+  * traditionally compiled dynamic indexing):
+  *
+  * 1. It has a default value of how a FIRRTL compiler compiles [[DontCare]].
+  *
+  * 2. The out-of-bounds behavior returns the zeroth element if the
+  *    [[SparseVec]] has a zeroth element.  Otherwise, it returns a
+  *    [[DontCare]].
+  *
+  * Note that this [[DontCare]] is likely not a true "don't care" that will be
+  * optimized to any value.  Instead, it is a value equal to how a FIRRTL
+  * compiler chooses to optimize a dynamic index into a wire vector initialized
+  * with a [[DontCare]].  This has historically been zero.
+  *
+  * A [[SparseVec]] will take up storage equal to the size of the provided
+  * mapping argument with one additional slot for the default value, if one is
+  * needed.
+  *
+  * @param size the apparent size of the vector
+  * @param gen the element type of the vector
+  * @param mapping the indices of the vector which are valid
+  */
+class SparseVec[A <: Data](size: Int, gen: => A, mapping: Seq[Int]) extends Record {
```

Comment:
It _can_ be calculated as the maximum of the mapping. @sequencer had the same suggestion. However, that prevents you from describing a dense vec-like thing that has the exact behavior you want. It could be a default like:

```scala
class SparseVec[A <: Data](mapping: Seq[Int])(size: Int = mapping.size, gen: => A) extends Record
```

This implementation is very much trying to just make this equivalent to the dense vec case, so it's likely not properly generalized. Do we want to generalize it right now?

WDYT?

Reply:
I think it's reasonable to just have the argument by default, we could have auxiliary constructors or factory methods to let the user derive the size from the mapping, eg.
```scala
  def this(gen: => A, mapping: Seq[Int]) = this(mapping.max, gen mapping)
```
Or similarly on an apply method on `object SparseVec`.

Reply:
Added alternative constructor in latest version.